### PR TITLE
Add dead letter topic support

### PR DESCRIPTION
**Pull request by yash-nisar**
Closes https://github.com/dapr/dotnet-sdk/issues/897

Signed-off-by: Yash Nisar <yashnisar@microsoft.com>

# Description
Add support for dead letter topic for the dotnet sdk

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: https://github.com/dapr/dotnet-sdk/issues/897

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [ ] Extended the documentation: Will do once I get approvals on the PR design

### CODE REVIEW

File: src/Dapr.AspNetCore/ITopicMetadata.cs [hunk 1]
```diff
@@ -28,6 +28,11 @@ public interface ITopicMetadata
         /// </summary>
         string PubsubName { get; }
 
+        /// <summary>
+        /// Gets the dead letter topic.
+        /// </summary>
+        string DeadLetterTopic { get; }
```
Comment (halspang) [inline src/Dapr.AspNetCore/ITopicMetadata.cs:34]: This is a breaking change: https://docs.microsoft.com/en-us/dotnet/core/compatibility/#members

The way we've done this in the past is usually adding a new interface that specifies what we're looking for. e.g. `IDeadLetterTopicMetadata`.

Truthfully though, I haven't really liked this pattern and I don't think it fits in with .NET's own recommendation for when to use interfaces. I'd try just adding the attribute directly to the `TopicAttribute` and see if that works. My only concern here is that the interfaces might be used for the encoding/decoding path.

File: src/Dapr.AspNetCore/TopicAttribute.cs [hunk 2]
```diff
@@ -105,6 +105,28 @@ public TopicAttribute(string pubsubName, string name, bool enableRawPayload, str
...
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TopicAttribute" /> class.
+        /// </summary>
+        /// <param name="pubsubName">The name of the pubsub component to use.</param>
+        /// <param name="name">The topic name.</param>
+        /// <param name="deadLetterTopic">The dead letter topic name.</param>
+        /// <param name="enableRawPayload">The enable/disable raw pay load flag.</param>
+        /// <param name="ownedMetadatas">The topic owned metadata ids.</param>
+        /// <param name="metadataSeparator">Separator to use for metadata.</param>
+        public TopicAttribute(string pubsubName, string name, string deadLetterTopic, bool enableRawPayload, string[] ownedMetadatas = null, string metadataSeparator = null)
```
Comment (halspang) [inline src/Dapr.AspNetCore/TopicAttribute.cs:117]: Were we not able to make a constructor with `TopicOptions`?

File: src/Dapr.AspNetCore/TopicOptions.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+// limitations under the License.
+// ------------------------------------------------------------------------
+
+namespace Dapr
+{
+    using System.Collections.Generic;
+    /// <summary>
+    /// This class defines configurations for the subscribe endpoint.
+    /// </summary>
+    public class TopicOptions
```
Comment (halspang) [inline src/Dapr.AspNetCore/TopicOptions.cs:20]: Is this missing Route and Priority?

File: src/Dapr.Client/Protos/dapr/proto/dapr/v1/appcallback.proto [hunk 4]
```diff
@@ -160,6 +160,9 @@ message TopicSubscription {
   // The optional routing rules to match against. In the gRPC interface, OnTopicEvent
   // is still invoked but the matching path is sent in the TopicEventRequest.
   TopicRoutes routes = 5;
+
+  // The optional dead letter queue for this topic to send events to.
+  string dead_letter_topic = 6;
```
Comment (halspang) [inline src/Dapr.Client/Protos/dapr/proto/dapr/v1/appcallback.proto:165]: Just to check, this is copied from `dapr/dapr`, right?

File: test/Dapr.AspNetCore.IntegrationTest.App/CustomTopicAttribute.cs [hunk 5]
```diff
@@ -30,5 +30,7 @@ public CustomTopicAttribute(string pubsubName, string name)
         public new string Match { get; }
 
         public int Priority { get; }
+
+        public string DeadLetterTopic { get; }
```
Comment (halspang) [inline test/Dapr.AspNetCore.IntegrationTest.App/CustomTopicAttribute.cs:34]: This change isn't required unless you want to use it in your test.

File: test/Dapr.AspNetCore.IntegrationTest/SubscribeEndpointTest.cs [hunk 6]
```diff
@@ -70,7 +78,7 @@ public async Task SubscribeEndpoint_ReportsTopics()
 
                         if (element.TryGetProperty("route", out JsonElement route))
                         {
-                            subscriptions.Add((pubsubName, topic, route.GetString(), rawPayload, string.Empty, originalMetadataString));
+                            subscriptions.Add((pubsubName, topic, route.GetString(), rawPayload, string.Empty, originalMetadataString, deadLetterTopic));
```
Comment (halspang) [inline test/Dapr.AspNetCore.IntegrationTest/SubscribeEndpointTest.cs:81]: Is this manually adding the DLQ to the subscription?

File: test/Dapr.AspNetCore.IntegrationTest/SubscribeEndpointTest.cs [hunk 7]
```diff
@@ -80,34 +88,35 @@ public async Task SubscribeEndpoint_ReportsTopics()
...
                                     var match = rule.GetProperty("match").GetString();
                                     var path = rule.GetProperty("path").GetString();
-                                    subscriptions.Add((pubsubName, topic, path, rawPayload, match, originalMetadataString));
+                                    subscriptions.Add((pubsubName, topic, path, rawPayload, match, originalMetadataString, deadLetterTopic));
                                 }
                             }
                             if (routes.TryGetProperty("default", out JsonElement defaultProperty))
                             {
-                                subscriptions.Add((pubsubName, topic, defaultProperty.GetString(), rawPayload, string.Empty, originalMetadataString));
+                                subscriptions.Add((pubsubName, topic, defaultProperty.GetString(), rawPayload, string.Empty, originalMetadataString, deadLetterTopic));
```
Comment (halspang) [inline test/Dapr.AspNetCore.IntegrationTest/SubscribeEndpointTest.cs:96]: Same question.

**Review by halspang — CHANGES_REQUESTED** [review]

File: examples/AspNetCore/RoutingSample/Startup.cs [hunk 8]
```diff
@@ -89,9 +90,20 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env, JsonSeri
...
 
+                var depositTopicOptions = new TopicOptions();
+                depositTopicOptions.PubsubName = PubsubName;
+                depositTopicOptions.Name = "deposit";
+                depositTopicOptions.DeadLetterTopic = "amountDeadLetterTopic";
+
+                var withdrawTopicOptions = new TopicOptions();
+                withdrawTopicOptions.PubsubName = PubsubName;
+                withdrawTopicOptions.Name = "withdraw";
+                withdrawTopicOptions.DeadLetterTopic = "amountDeadLetterTopic";
```
[the remark this answers is not in the document]
Comment (halspang) [reply examples/AspNetCore/RoutingSample/Startup.cs:101]: Just realized this was in the `RoutingSample` and not the `ControllerSample`. Nevermind.

File: src/Dapr.AspNetCore/DaprEndpointRouteBuilderExtensions.cs [hunk 9]
```diff
@@ -62,7 +62,7 @@ private static IEndpointConventionBuilder CreateSubscribeEndPoint(IEndpointRoute
 
             return endpoints.MapGet("dapr/subscribe", async context =>
             {
-                var logger = context.RequestServices.GetService<ILoggerFactory>().CreateLogger("DaprTopicSubscription");
+                var logger = context.RequestServices.GetService<ILoggerFactory>().CreateLogger("DaprTopicSubscription");
```
[the remark this answers is not in the document]
Comment (yash-nisar) [reply src/Dapr.AspNetCore/DaprEndpointRouteBuilderExtensions.cs:65]: Fixed

File: examples/AspNetCore/ControllerSample/Controllers/SampleController.cs [hunk 10]
```diff
@@ -64,38 +64,65 @@ public ActionResult<Account> Get([FromState(StoreName)] StateEntry<Account> acco
         /// <param name="daprClient">State client to interact with Dapr runtime.</param>
         /// <returns>A <see cref="Task{TResult}"/> representing the result of the asynchronous operation.</returns>
         ///  "pubsub", the first parameter into the Topic attribute, is name of the default pub/sub configured by the Dapr CLI.
-        [Topic("pubsub", "deposit")]
+        [Topic("pubsub", "deposit", "amountDeadLetterTopic", false)]
         [HttpPost("deposit")]
         public async Task<ActionResult<Account>> Deposit(Transaction transaction, [FromServices] DaprClient daprClient)
         {
-            logger.LogDebug("Enter deposit");
+            Console.WriteLine("Enter the amount to be deposited: ");
```
[the remark this answers is not in the document]
Comment (yash-nisar) [reply examples/AspNetCore/ControllerSample/Controllers/SampleController.cs:71]: Fixed

File: examples/AspNetCore/RoutingSample/Startup.cs [hunk 11]
```diff
@@ -1,4 +1,4 @@
-// ------------------------------------------------------------------------
+﻿// ------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (yash-nisar) [reply examples/AspNetCore/RoutingSample/Startup.cs:1]: I'll probably add a small section to test the dead letter topic by providing a negative amount to be deposited/withdrawn

File: src/Dapr.AspNetCore/DaprEndpointConventionBuilderExtensions.cs [hunk 12]
```diff
@@ -103,5 +103,39 @@ public static T WithTopic<T>(this T builder, string pubsubName, string name, boo
...
+                throw new ArgumentNullException(nameof(builder));
+            }
+
+            ArgumentVerifier.ThrowIfNullOrEmpty(topicOptions.PubsubName, nameof(topicOptions.PubsubName));
+            ArgumentVerifier.ThrowIfNullOrEmpty(topicOptions.Name, nameof(topicOptions.Name));
+
+            var topicObject = new TopicAttribute(topicOptions.PubsubName, topicOptions.Name,
+                topicOptions.DeadLetterTopic, topicOptions.EnableRawPayload);
+
+            builder.WithMetadata(topicObject);
```
[the remark this answers is not in the document]
Comment (yash-nisar) [reply src/Dapr.AspNetCore/DaprEndpointConventionBuilderExtensions.cs:128]: I just copied the concept from the previous method: https://github.com/dapr/dotnet-sdk/blob/master/src/Dapr.AspNetCore/DaprEndpointConventionBuilderExtensions.cs#L96

File: src/Dapr.AspNetCore/TopicOptions.cs (hunk 3 above)
Comment (yash-nisar) [reply src/Dapr.AspNetCore/TopicOptions.cs:20]: Fixed

File: src/Dapr.Client/Protos/dapr/proto/dapr/v1/appcallback.proto (hunk 4 above)
Comment (yash-nisar) [reply src/Dapr.Client/Protos/dapr/proto/dapr/v1/appcallback.proto:165]: Yes

File: test/Dapr.AspNetCore.IntegrationTest.App/CustomTopicAttribute.cs (hunk 5 above)
Comment (yash-nisar) [reply test/Dapr.AspNetCore.IntegrationTest.App/CustomTopicAttribute.cs:34]: I had to add it because I had added `DeadLetterTopic` to `ITopicMetadata`.

File: test/Dapr.AspNetCore.IntegrationTest/SubscribeEndpointTest.cs (hunk 6 above)
Comment (yash-nisar) [reply test/Dapr.AspNetCore.IntegrationTest/SubscribeEndpointTest.cs:81]: Discussed offline

File: test/Dapr.AspNetCore.IntegrationTest/SubscribeEndpointTest.cs (hunk 7 above)
Comment (yash-nisar) [reply test/Dapr.AspNetCore.IntegrationTest/SubscribeEndpointTest.cs:96]: Discussed offline

File: src/Dapr.AspNetCore/ITopicMetadata.cs (hunk 1 above)
Comment (yash-nisar) [reply src/Dapr.AspNetCore/ITopicMetadata.cs:34]: Fixed !

File: src/Dapr.AspNetCore/TopicAttribute.cs (hunk 2 above)
Comment (yash-nisar) [reply src/Dapr.AspNetCore/TopicAttribute.cs:117]: Discussed offline

File: examples/AspNetCore/ControllerSample/Controllers/SampleController.cs [hunk 13]
```diff
@@ -64,38 +64,65 @@ public ActionResult<Account> Get([FromState(StoreName)] StateEntry<Account> acco
         /// <param name="daprClient">State client to interact with Dapr runtime.</param>
         /// <returns>A <see cref="Task{TResult}"/> representing the result of the asynchronous operation.</returns>
         ///  "pubsub", the first parameter into the Topic attribute, is name of the default pub/sub configured by the Dapr CLI.
-        [Topic("pubsub", "deposit")]
+        [Topic("pubsub", "deposit", "amountDeadLetterTopic", false)]
         [HttpPost("deposit")]
         public async Task<ActionResult<Account>> Deposit(Transaction transaction, [FromServices] DaprClient daprClient)
         {
-            logger.LogDebug("Enter deposit");
+            Console.WriteLine("Enter deposit method...");
```
Comment (halspang) [inline examples/AspNetCore/ControllerSample/Controllers/SampleController.cs:71]: We should still be using the logger.

File: src/Dapr.AspNetCore/TopicOptions.cs [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+        /// </summary>
+        public int Priority { get; set; }
+
+        /// <summary>
+        /// Gets or Sets the <see cref="IOriginalTopicMetadata.Id"/> owned by topic.
+        /// </summary>
+        public string[] OwnedMetadatas { get; set; }
+
+        /// <summary>
+        /// Get or Sets the separator to use for metadata
```
Comment (halspang) [inline src/Dapr.AspNetCore/TopicOptions.cs:53]: Nit, first comment should end in a `.`.

**Review by halspang — COMMENTED** [review]
Just some minor feedback and you're good to go.

File: examples/AspNetCore/ControllerSample/Controllers/SampleController.cs (hunk 13 above)
Comment (yash-nisar) [reply examples/AspNetCore/ControllerSample/Controllers/SampleController.cs]: Fixed

File: src/Dapr.AspNetCore/TopicOptions.cs (hunk 14 above)
Comment (yash-nisar) [reply src/Dapr.AspNetCore/TopicOptions.cs]: Fixed

**Review by halspang — APPROVED** [review]
